### ISSUE TITLE
[Idea] Documentation Catalogues

### ISSUE THREAD

**Issue by clsource** [onyx-lang/onyx#167]
Hello since _Onyx_ have proper documentation comments in the parser, I think it can use documentation catalogues similar to [Swift Docc](https://www.swift.org/documentation/docc/) tool.

I tried to create something for _Odin_, but since it does not support documentation comments (only normal comments with special structure), it proved to be quite difficult without parser modification.

But I learned alot about how to structure documentation catalogues.

You can check an example output here (that was manually created to test the structure)

https://odin-arsenal.github.io/gungnir/SlothCreator/

A documentation catalogue consist of this structure:

- A directory where all the content will be stored (markdown files and resources)
- The `mdbook` and configuration files for generating the output.
- An `articles` directory to store different complementary documentation.
- A `tutorials` directory  to store step by step instructions on how to use the packages and apis.
- A `extensions` directory. Is a special directory that follows the same project structure. Can overwrite the documentation for any file, method, structure, enum, constant. The tool would read this directory and use these files for rendering the markdown instead of the contents of doc comments.

Normally it will have a structure similar to this:

```text
.
├── LICENSE.md
├── README.md
├── book.toml
├── preprocessors
│   └── set_version.py
├── run
├── src
│   ├── README.md
│   ├── SUMMARY.md
│   ├── articles
│   ├── extensions
│   │   └── core
│   │       ├── (collection)overview.md
│   │       └── slice
│   │           ├── (package)overview.md
│   │           └── slice.odin.d
│   │               ├── (file)overview.md
│   │               ├── examples
│   │               │   ├── reduce-example1
│   │               │   │   └── example.odin.txt
│   │               │   └── reduce_reverse-example1
│   │               │       └── example.odin.txt
│   │               ├── reduce.md
│   │               └── reduce_reverse.md
│   ├── resources
│   │   └── odin.jpg
│   └── tutorials
└── theme
    ├── book.js
    └── highlight.js
```

The current documentation tool https://github.com/onyx-lang/onyx-doc-gen is awesome to create API documentation, but it have some disadvantages:

1. Can only create HTML with a specified template. By using markdown it can be converted to other formats (using pandoc) and create other artifacts such as PDF or Epub. Even a tar file can be created to easily send over the network documentation and open them easily.

2. The current implementation is limited to docs inside the code. Extending the documentation with complementary articles and tutorials is an important tool to improve adoption of the packages and well usage of APIS.

3. Too much different contexts on a single page. The separation in a single page per context can improve the reader experience when looking for symbols, using the `mdbook` tools we can create confortable documentation that does not overwhelm the reader.

4. Separation of concerns, the core apis are mixed with the user package apis. It would be a good idea to generate documentation per package, that way it would be easier to version each component.

**Comment by brendanfh** [comment]
I completely agree that a more comprehensive documentation generation tool is needed for Onyx. The current documentation generation was the best I could make with how the compiler was structured at the time. I think with some changes I am working on right now in the toolchain, it will be much easier to create a separate tool that can do this generation.

I am also very interested in the Doctests that you mentioned. I think they will be relatively easy to implement while creating the new documentation generation tool.

All around I think this would be a great addition to the Onyx toolchain.

**Comment by clsource** [comment]
I implemented SlothCreator example in Onyx for testing

[SlothCreator.zip](https://github.com/user-attachments/files/17969469/SlothCreator.zip)

**Comment by clsource** [comment]
A related link for developer docs: https://diataxis.fr/ 

Diátaxis is a way of thinking about and doing documentation.
tutorials, how-to guides, technical reference and explanation. It places them in a systematic relationship, and proposes that documentation should itself be organised around the structures of those needs.
